packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/HeatmapsControllerTest.java

```diff
@@ -0,0 +1,104 @@
+package io.flutter.plugins.googlemaps;
+
+import static io.flutter.plugins.googlemaps.Convert.HEATMAP_DATA_KEY;
+import static io.flutter.plugins.googlemaps.Convert.HEATMAP_ID_KEY;
+import static io.flutter.plugins.googlemaps.Convert.HEATMAP_OPACITY_KEY;
+import static io.flutter.plugins.googlemaps.Convert.weightedDataToJson;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.when;
+
+import android.os.Build;
+import com.google.android.gms.maps.GoogleMap;
+import com.google.android.gms.maps.model.LatLng;
+import com.google.android.gms.maps.model.TileOverlay;
+import com.google.android.gms.maps.model.TileOverlayOptions;
+import com.google.maps.android.heatmaps.HeatmapTileProvider;
+import com.google.maps.android.heatmaps.WeightedLatLng;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentMatcher;
+import org.mockito.Mockito;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(sdk = Build.VERSION_CODES.P)
+public class HeatmapsControllerTest {
+  private HeatmapsController controller;
+  private GoogleMap googleMap;
+
+  @Before
+  public void setUp() {
+    controller = spy(new HeatmapsController());
+    googleMap = mock(GoogleMap.class);
+    controller.setGoogleMap(googleMap);
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void controller_AddHeatmapThrowsErrorIfHeatmapIdIsNull() {
+    final Map<String, String> heatmapOptions = new HashMap<>();
+
+    final List<Object> heatmaps = Collections.singletonList(heatmapOptions);
+    try {
+      controller.addHeatmaps(heatmaps);
+    } catch (IllegalArgumentException e) {
+      assertEquals("heatmapId was null", e.getMessage());
+      throw e;
+    }
+  }
+
+  @Test
+  public void controller_AddChangeAndRemoveHeatmap() {
```

Comment:
Because I based this test off of `MarkersControllerTest.java` which does the same

Reply:
I'm generally not a fan of one test doing multiple things, but I guess in practice if we broke this up it would be add, add/change, and add/remove, so maybe more duplication than it's worth.